File: ios/SecureCore/Sources/SecureCore/KeyManager/WrapsEnvelope.swift
import Foundation

/// Canonical envelope for a wrapped DEK.
///
/// See `docs/wraps-schema-v1.md` for the frozen specification.
public struct WrapsEnvelope: Codable, Equatable {
    public static let currentSchemaVersion = "1.1"
    public static let algoKeychainGCM = "AES-256-GCM-KEYCHAIN"

    public let schemaVersion: String
    public let device: DeviceWrap?
    public let recovery: AnyCodable?

    enum CodingKeys: String, CodingKey {
        case schemaVersion = "schema_version"
        case device
        case recovery
    }

    public init(schemaVersion: String, device: DeviceWrap?, recovery: AnyCodable? = nil) {
        self.schemaVersion = schemaVersion
        self.device = device
        self.recovery = recovery
    }

    public static func fromJson(_ data: Data) throws -> WrapsEnvelope {
        let decoder = JSONDecoder()
        let envelope: WrapsEnvelope
        do {
            envelope = try decoder.decode(WrapsEnvelope.self, from: data)
        } catch {
            throw KeyManagerError.invalidWrapsFormat("invalid JSON: \(error.localizedDescription)")
        }

        if envelope.schemaVersion.isEmpty {
            throw KeyManagerError.invalidWrapsFormat("missing schema_version")
        }
        if envelope.schemaVersion != currentSchemaVersion {
            throw KeyManagerError.versionTooNew(
                found: envelope.schemaVersion, supported: currentSchemaVersion)
        }

        return envelope
    }

    public func toJson() throws -> Data {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        return try encoder.encode(self)
    }

    public func validate() throws {
        if schemaVersion != Self.currentSchemaVersion {
            throw KeyManagerError.versionTooNew(
                found: schemaVersion, supported: Self.currentSchemaVersion)
        }

        guard let device = device else {
            throw KeyManagerError.invalidWrapsFormat("device must not be null")
        }

        if device.algo.isEmpty {
            throw KeyManagerError.invalidWrapsFormat("device.algo must not be empty")
        }
        if device.keyAlias.isEmpty {
            throw KeyManagerError.invalidWrapsFormat("device.key_alias must not be empty")
        }

        let ivBytes = try device.ivData()
        if ivBytes.count != 12 {
            throw KeyManagerError.invalidWrapsFormat(
                "device.iv must be 12 bytes, got \(ivBytes.count)")
        }

        let tagBytes = try device.tagData()
        if tagBytes.count != 16 {
            throw KeyManagerError.invalidWrapsFormat(
                "device.tag must be 16 bytes, got \(tagBytes.count)")
        }

        let ciphertextBytes = try device.ciphertextData()
        if ciphertextBytes.isEmpty {
            throw KeyManagerError.invalidWrapsFormat("device.ciphertext must not be empty")
        }
    }
}

public struct DeviceWrap: Codable, Equatable {
    public let algo: String
    public let keyAlias: String
    public let iv: String
    public let tag: String
    public let ciphertext: String

    enum CodingKeys: String, CodingKey {
        case algo
        case keyAlias = "key_alias"
        case iv
        case tag
        case ciphertext
    }

    public init(algo: String, keyAlias: String, iv: String, tag: String, ciphertext: String) {
        self.algo = algo
        self.keyAlias = keyAlias
        self.iv = iv
        self.tag = tag
        self.ciphertext = ciphertext
    }

    public func ivData() throws -> Data {
        guard let data = Data(base64Encoded: iv) else {
            throw KeyManagerError.invalidWrapsFormat("device.iv: invalid base64")
        }
        return data
    }

    public func tagData() throws -> Data {
        guard let data = Data(base64Encoded: tag) else {
            throw KeyManagerError.invalidWrapsFormat("device.tag: invalid base64")
        }
        return data
    }

    public func ciphertextData() throws -> Data {
        guard let data = Data(base64Encoded: ciphertext) else {
            throw KeyManagerError.invalidWrapsFormat("device.ciphertext: invalid base64")
        }
        return data
    }
}

/// Minimal type-erased Codable wrapper for the `recovery` field.
public struct AnyCodable: Codable, Equatable {
    public let value: [String: String]

    public init(_ value: [String: String]) {
        self.value = value
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if container.decodeNil() {
            value = [:]
        } else {
            value = try container.decode([String: String].self)
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        if value.isEmpty {
            try container.encodeNil()
        } else {
            try container.encode(value)
        }
    }
}

File: android/secure-core-android/src/main/kotlin/com/securecore/keymanager/WrapsEnvelope.kt
package com.securecore.keymanager

import android.util.Base64
import org.json.JSONObject

/**
 * Canonical envelope for a wrapped DEK.
 *
 * See `docs/wraps-schema-v1.md` for the frozen specification.
 */
data class WrapsEnvelope(
    val schemaVersion: String,
    val device: DeviceWrap?,
    val recovery: JSONObject? = null
) {
    companion object {
        const val CURRENT_SCHEMA_VERSION = "1.1"
        const val ALGO_AES_256_GCM_KEYSTORE = "AES-256-GCM-KEYSTORE"

        /**
         * Parses a [WrapsEnvelope] from its JSON string representation.
         *
         * @throws KeyManagerError.WrapFormatInvalid if JSON is malformed or missing required fields.
         * @throws KeyManagerError.VersionTooNew if schema_version is unsupported.
         */
        fun fromJson(json: String): WrapsEnvelope {
            val obj = try {
                JSONObject(json)
            } catch (e: Exception) {
                throw KeyManagerError.WrapFormatInvalid("invalid JSON: ${e.message}")
            }

            val version = obj.optString("schema_version", "")
            if (version.isEmpty()) {
                throw KeyManagerError.WrapFormatInvalid("missing schema_version")
            }
            if (version != CURRENT_SCHEMA_VERSION) {
                throw KeyManagerError.VersionTooNew(version, CURRENT_SCHEMA_VERSION)
            }

            val deviceObj = if (obj.isNull("device")) null else obj.optJSONObject("device")
            val device = deviceObj?.let { d ->
                DeviceWrap(
                    algo = d.optString("algo", "").also {
                        if (it.isEmpty()) throw KeyManagerError.WrapFormatInvalid("missing device.algo")
                    },
                    keyAlias = d.optString("key_alias", "").also {
                        if (it.isEmpty()) throw KeyManagerError.WrapFormatInvalid("missing device.key_alias")
                    },
                    iv = d.optString("iv", "").also {
                        if (it.isEmpty()) throw KeyManagerError.WrapFormatInvalid("missing device.iv")
                    },
                    tag = d.optString("tag", "").also {
                        if (it.isEmpty()) throw KeyManagerError.WrapFormatInvalid("missing device.tag")
                    },
                    ciphertext = d.optString("ciphertext", "").also {
                        if (it.isEmpty()) throw KeyManagerError.WrapFormatInvalid("missing device.ciphertext")
                    }
                )
            }

            val recovery = if (obj.isNull("recovery")) null else obj.optJSONObject("recovery")

            return WrapsEnvelope(
                schemaVersion = version,
                device = device,
                recovery = recovery
            )
        }
    }

    /**
     * Serializes this envelope to its canonical JSON string.
     */
    fun toJson(): String {
        val obj = JSONObject()
        obj.put("schema_version", schemaVersion)

        if (device != null) {
            val d = JSONObject()
            d.put("algo", device.algo)
            d.put("key_alias", device.keyAlias)
            d.put("iv", device.iv)
            d.put("tag", device.tag)
            d.put("ciphertext", device.ciphertext)
            obj.put("device", d)
        } else {
            obj.put("device", JSONObject.NULL)
        }

        if (recovery != null) {
            obj.put("recovery", recovery)
        } else {
            obj.put("recovery", JSONObject.NULL)
        }

        return obj.toString()
    }

    /**
     * Validates this envelope. Throws on error.
     *
     * @throws KeyManagerError.WrapFormatInvalid if device is null or fields are invalid.
     * @throws KeyManagerError.VersionTooNew if schema_version is unsupported.
     * @throws KeyManagerError.AlgoUnsupported if algo is not recognized.
     */
    fun validate() {
        if (schemaVersion != CURRENT_SCHEMA_VERSION) {
            throw KeyManagerError.VersionTooNew(schemaVersion, CURRENT_SCHEMA_VERSION)
        }
        val d = device ?: throw KeyManagerError.WrapFormatInvalid("device must not be null")
        if (d.algo.isEmpty()) throw KeyManagerError.WrapFormatInvalid("device.algo must not be empty")
        if (d.keyAlias.isEmpty()) throw KeyManagerError.WrapFormatInvalid("device.key_alias must not be empty")

        val ivBytes = d.ivBytes()
        if (ivBytes.size != 12) {
            throw KeyManagerError.WrapFormatInvalid("device.iv must be 12 bytes, got ${ivBytes.size}")
        }
        val tagBytes = d.tagBytes()
        if (tagBytes.size != 16) {
            throw KeyManagerError.WrapFormatInvalid("device.tag must be 16 bytes, got ${tagBytes.size}")
        }
        if (d.ciphertextBytes().isEmpty()) {
            throw KeyManagerError.WrapFormatInvalid("device.ciphertext must not be empty")
        }

        if (d.algo != ALGO_AES_256_GCM_KEYSTORE) {
            throw KeyManagerError.AlgoUnsupported(d.algo)
        }
    }
}

/**
 * Device-local DEK wrap produced by the OS keystore.
 */
data class DeviceWrap(
    val algo: String,
    val keyAlias: String,
    val iv: String,
    val tag: String,
    val ciphertext: String
) {
    fun ivBytes(): ByteArray = Base64.decode(iv, Base64.DEFAULT)
    fun tagBytes(): ByteArray = Base64.decode(tag, Base64.DEFAULT)
    fun ciphertextBytes(): ByteArray = Base64.decode(ciphertext, Base64.DEFAULT)
}
